experimental/ecma402_listformat/src/lib.rs

```diff
@@ -0,0 +1,565 @@
+//! assert_eq!(
+//!     vec!["a", "b"],
+//!     owned.variants()
+//!         .map(|v| v.to_owned())
+//!         .collect::<Vec<String>>(),
+//! );
+//!
+//! ```
+//!
+
+/// This trait contains the common features of the Locale object that must be shared among
+/// all the implementations.  Every implementor of `listformat` should provide their
+/// own version of [Locale], and should ensure that it implements [Locale]. as
+/// specified here.
+///
+/// A stronger type definition is also possible, pending agreement.  Until no firm implementations
+/// exist, we reserve the option to strenghten the contract of [Locale] as we see fit.
+///
+/// For example, here are some possible methods that this trait could implement instead:
+///
+/// - `language`
+/// - `script`
+/// - `variant`
+/// - (...)
+pub trait Locale<'a> {
+    /// Gives the full representation of the `Locale` in the [BCP 47][bcp47] string encoding.
+    ///
+    ///   [bcp47]: https://tools.ietf.org/html/bcp47
+    fn to_bcp47() -> &'a str;
+}
+
+/// The [listformat] mod contains all the needed implementation bits for `Intl.ListFormat`.
+///
+/// > Note: This is not yet the entire API.  I'd like to get to a consensus on what has been
+/// defined, then use the patterns adopted here for the rest.
+pub mod listformat {
+    /// Contains the API configuration as prescribed by ECMA 402.
+    ///
+    /// The meaning of the options is the same as in the similarly named
+    /// options in the JS version.
+    ///
+    /// See [options::In] for the contents of the options.  See the [Factory]
+    /// trait for the use of the options.
+    pub mod options {
+        /// Chooses the list formatting approach.
+        #[derive(Eq, PartialEq, Debug, Clone)]
+        pub enum Style {
+            Long,
+            Short,
+            Narrow,
+        }
+        /// Chooses between "this, that and other", and "this, that or other".
+        #[derive(Eq, PartialEq, Debug, Clone)]
+        pub enum Type {
+            /// "This, that and other".
+            Conjunction,
+            /// "This, that or other".
+            Disjunction,
+        }
+
+        /// The options set by the user at construction time.  See discussion at the top level
+        /// about the name choice.  Provides as a "bag of options" since we don't expect any
+        /// implementations to be attached to this struct.
+        #[derive(Eq, PartialEq, Debug, Clone)]
+        pub struct In {
+            /// Selects a [Style] for the formatted list.
+            pub style: Style,
+            /// Selects a [Type] for the formatted list.
+            pub in_type: Type,
+        }
+    }
+
+    /// A factory object containing the methods used to construct new [Format], as well
+    /// as associated functions prescribed by `Intl.ListFormat`.
+    ///
+    /// While ECMA 402 originally has functions under `Intl`, we probably want to
+    /// obtain a separate factory from each implementor.
+    pub trait Factory<'a> {
```

Comment:
I don't understand the separation between `Factory` and `Format`.

Reply:
In Rust we usually try to localize bounds, including lifetimes. Instead of setting `<'a>` on the trait, you should have that on methods that need it. `try_new` should not need it because we don't pass any reference to be stored on the struct (I mentioned above that `Locale` shouldn't have lifetime).

Reply:
Re `Factory`: If we wanted to ensure any implementation has a `supportedLocalesOf`, how'd we do it?

Re bounds: done.

Reply:
We agreed to omit `supported_locales_of`.  It is more in the spirit of Rust to factor locale negotiation out of this functionality.